source/agora/test/EnrollmentManager.d

```diff
@@ -221,3 +221,188 @@ unittest
+            Ledger ledger, TaskManager taskman)
+        {
+            this.ledger = ledger;
+            super(network, key_pair, ledger, taskman);
+        }
+
+        ///
+        public override bool isNominating () @safe @nogc nothrow
+        {
+            return this.is_nominating;
+        }
+
+        ///
+        public override void tryNominate () @safe
+        {
+            if (this.ledger.getBlockHeight < 1)
+            {
+                super.tryNominate();
+            }
+        }
+    }
+
+    static class OneTimeValidator : TestValidatorNode
+    {
+        /// Ctor
+        public this (Config config, Registry* reg, immutable(Block)[] blocks)
+        {
+            super(config, reg, blocks);
+        }
+
+        ///
+        protected override Nominator getNominator (
+            NetworkManager network, KeyPair key_pair, Ledger ledger,
+            TaskManager taskman)
+        {
+            return new OneTimeNominator(network, key_pair, ledger, taskman);
+        }
+    }
+
+    static class BadAPIManager : TestAPIManager
+    {
+        ///
+        public this (immutable(Block)[] blocks)
+        {
+            super(blocks);
+        }
+
+        /// see base class
+        public override void createNewNode (Config conf)
+        {
+            if (this.nodes.length == 0)
+            {
+                auto api = RemoteAPI!TestAPI.spawn!MisbehavingValidator(
+                    conf, &this.reg, this.blocks, conf.node.timeout);
+                this.reg.register(conf.node.address, api.tid());
+                this.nodes ~= NodePair(conf.node.address, api);
+            }
+            else
+            {
+                auto api = RemoteAPI!TestAPI.spawn!OneTimeValidator(
+                    conf, &this.reg, this.blocks, conf.node.timeout);
+                this.reg.register(conf.node.address, api.tid());
+                this.nodes ~= NodePair(conf.node.address, api);
+            }
+        }
+    }
+
+    auto network = makeTestNetwork!BadAPIManager(TestConf.init);
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.start();
+    network.waitForDiscovery();
+    auto nodes = network.clients;
+    auto bad_validator = nodes[0];
+    auto validator = nodes[1];
+
+    // Sanity check: Check if genesis block has enrollments
+    const b0 = validator.getBlocksFrom(0, 2)[0];
+    assert(b0.header.enrollments.length >= 1);
+
+    // Make a bloack using Genesis
+    Transaction[] txs;
+    txs = genesisSpendable().take(Block.TxsInBlock).enumerate()
+        .map!(en => en.value.refund(WK.Keys.A.address).sign())
+        .array();
+    txs.each!(tx => validator.putTransaction(tx));
+
+    // Ensure everyone is at the same level
+    ensureConsistency(nodes, 1);
+
+    // Make a block in the `bad_validator`
+    txs = txs
+        .map!(ptx => TxBuilder(ptx).refund(WK.Keys.B.address).sign())
+        .array();
+    txs.each!(tx => bad_validator.putTransaction(tx));
```

Comment:
There's nothing special about the block being created by `bad_validator` though.
Also, the BadValidator should perhaps have a boolean which shows the code was executed. I will help you with this.

Reply:
> There's nothing special about the block being created by bad_validator though.
Ah, I know what you mean. The bad validator only nominates a block with bad data although it makes valid transactions.

> Also, the BadValidator should perhaps have a boolean which shows the code was executed. I will help you with this.
Yeah, thanks in advance.